# Stacked taxonomy bar graphs 
<!-- Chunk to allow horizontal scroll in chunks rather than wrap text  -->
```{css, echo=FALSE}
pre, code {white-space:pre !important; overflow-x:auto}
```

One common way to begin analyzing your data is with stacked taxonomy bar graph. 
Theses graphs are great for looking at the distribution of different taxa across all of your samples at the same time.
Yet, keep in mind that they are best used for simpler data given that there is only about a dozen or so of different colors that are easily distinguishable by the human eye. 
Still, even tho no one likes stacked taxonomy  bar graphs, they have somehow found a way into microbial ecology, so we as humans just keep making them.  

Before beginning, it is important to consider you own data and what would be the best way to present your figures to the readers. 
In the case of the data used in this pipeline, we had **two distinct experiences**, 
our samples were either **"raw" groundwater** samples collected and filtered on the same day, or **rock pellets** and **groundwater** samples collected at **different times** from **three different bioreactors**... so lots of different plots to generate with different variables...  

In the end, we decided to present our data in the following way which we deemed the easiest to interpret by the readers : 

```{r echo=FALSE, out.width = "90%", fig.align = "center", out.lenght = "90%"}
knitr::include_graphics("data/Experience 1 - Groundwater-family.png")
```

```{r echo=FALSE, out.width = "90%", fig.align = "center", out.lenght = "90%"}
knitr::include_graphics("data/Experience 2 - Groundwater-family.png")
```

Below you will find all the lines of codes which were used to generate these figures along with a thorough description of the code to allow you to adapt this code to your own data. 
We believe this code to be attractive to others based on the following characteristics : 

- The taxonomy rank of interest is defined only once (to avoid having to find and replace all occurrence of the taxonomic rank to adapt this scrip to different ranks of interest)
- The use of a *for loop* to generate and save multiple figures based on a defined iterator ([for more details on iterations and iterators](https://r4ds.had.co.nz/iteration.html))
- The grouping of less abundant taxa in the category *Others*
- The fact that this grouping is done for every sample individually using a *for loop*
- Each taxa is assigned a specific color and this same color is used in the different figures 
- The names of the taxa are presented in italic 
- The most abundant taxa is aligned along the bottom of the chart to allow a better comparison between samples
- The legend is presented in alphabetical order with the exception of the category *Others* which figures at the top 

Finally, the last section of this guide describes further ways of manipulating and presenting your plots to generate beautiful and publishable figures. 

## Required libraries 
### Installing libraries
These libraries were installed in R version 4.3.1 
```{r, eval=FALSE, cache=TRUE}
install.packages("devtools")
devtools::install_github("andyofsmeg/ggTag")
install.packages("forcats")
install.packages("dplyr")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.17")
BiocManager::install('phyloseq')
install.packages("randomcoloR")
install.packages("ggplot2")
install.packages("tidyr")
install.packages("stringr")
install.packages("ggtext")
install.packages("DT")
```
### Loading the libraries
```{r, message=FALSE, warning=FALSE, cache=TRUE, eval=FALSE}
library(devtools)
library(ggTag) # To extract title from plots
library(forcats) # To reorder our factors (function "fct_relevel")
library(dplyr) # Dataframe manipulation (notably function "pull")
library(tidyr) # Dataframe manipulation (function "separate")
library(phyloseq) # Very pratical library for the analysis of amplicon data 
library(randomcoloR) # Generate sets of random colors
library(ggplot2) # Generate plots 
library(stringr) # Makes working with strings as easy as possible (function "str_replace")
library(ggtext) # Allows the use of markdown text (used to make names italic)
library(glue)
```


## Required files

Three files are required for this script : 

1. Amplicon sequence variant (ASV) table (`rarefied_asv_bac.csv`)
2. Taxonomy table (`rarefied_taxa_bac.csv`)
3. Sample metadata table (`rarefied_meta_bac.csv`)

**ASV and Taxonomy table **

In our case, the first two files were generated from Illumina-sequenced paired-end fastq files using the [DADA2 pipeline v1.16](https://benjjneb.github.io/dada2/tutorial.html). 
To control for uneven sequencing effort in amplicon sequence analyses we normalized the sampling depth by the random subsampling of sequences from each sample down to the lowest but reasonable sample’s depth. 
This normalization method is refereed to as rarefying. While this approach is the subject of considerable debate and statistical criticism (see the 2014 PLOS Computational Biology paper, “[Waste not, want not: why rarefying microbiome data is inadmissible](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003531)” by McMurdie and Holmes) and alternative methods have been developed ([DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), [cumulative sum scaling (CSS)](https://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html), and more...) rarefaction is still widely used and very popular in the literature. **The script used for the rarefaction of our sample is available [here](insert script).** 

**Sample metadata table**

Metadata provides the key to gaining biological insight from your data. The compiling of sample metadata is typically a step you will have started before beginning your amplicon sequence analysis. 
The metadata table typically contains descriptions of the samples, such as origin, sample type, geophysical characteristics, time point, etc... 

```{r, cache=TRUE, eval=FALSE}
path="~/project/domain/int_data"

asv=read.table(file = glue("{path}/rarefied_asv_bac.csv"), sep=",", row.names=1, header=TRUE, check.names=FALSE)
taxa=read.table(file = glue("{path}/rarefied_taxa_bac.csv"), sep=",", row.names=1, header=TRUE)
meta=read.table(file = glue("{path}/rarefied_meta_bac.csv"), sep=",", row.names=1, header=TRUE)
```

```{r, echo=FALSE, cache=TRUE, eval=FALSE}
#Remove unwanted sample and column from the sample metadata file 
meta=subset(meta, !Sample_type=='Peatbog') # Remove sample from the peatbog
meta=subset(meta, select=-c(is.neg))
meta$position_pastille[meta$Sample_type == "Groundwater"] <- NA
```

Below is a preview of these three table.

```{r, echo=FALSE}
library(glue)
path="/Users/karinevilleneuve/Library/CloudStorage/OneDrive-UQAM/Github/amplicon-analysis-guide/data"
asv=read.table(file = glue("{path}/rarefied_asv_bac.csv"), sep=",", row.names=1, header=TRUE, check.names=FALSE)
taxa=read.table(file = glue("{path}/rarefied_taxa_bac.csv"), sep=",", row.names=1, header=TRUE)
meta=read.table(file = glue("{path}/rarefied_meta_bac.csv"), sep=",", row.names=1, header=TRUE)
```

```{r, echo=FALSE, cache=TRUE}
library(DT)
datatable(head(asv[1:6,6:9]), caption="ASV table", options=list(scrollX=T, dom = 't', paging=FALSE, ordering=FALSE,
                                                            initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '12px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '12px')

datatable(head(taxa), caption="Taxonomy table", options=list(scrollX=T, dom = 't', paging=FALSE, ordering=FALSE, 
                                                            initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '12px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '12px')

datatable(meta, caption="Sample metadata table", options=list(dom = 'tip', pageLength=5,autoWidth = TRUE,columnDefs=list(list(width="50%", targets=0)),
                                                              initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '12px'});",
                                                                              "}"))) %>% formatStyle(columns = colnames(.$x$data), `font-size` = '12px')
``` 

## Defining taxonomic rank

This is were we define which taxonomy rank we want to show in our figures. In this example we want to display the relative abundance of taxon at the genus level. 
```{r, cache=TRUE, eval=FALSE}
taxa_rank="Genus"
```

## Prepping the metadata table 

Getting your data ready for the analysis may not be the longest and most complex part of your code (especially compared to plotting the figures...), but may be the part where you spend a lot of time thinking about your own set of data. I recommend visualizing and even drawing different sets of figures in order to help you identify which variables to consider in your own code. As a reminder... 

***

<p style="text-align: center;">“*Data Scientists spend up to 80% of the time on data cleaning and 20% of their time on actual data analysis*”</p>
<p style="text-align: right;font-size:8pt">Dasu and Johnson, 2003</p>
***

In our case, we first had to generate new columns in the sample metadata table with some crucial information (**experience #**, **time**, **bioreactor #**, and **replicate #**) which were only defined in the sample names. This first chunk of code was therefore used to extract the **"row.names**" from the `meta` dataframe into a new column named **"sample_name"**, which is then split into five new columns using the hyphen (-) as separator. The last three lines of code were used to replace certain abbreviation and words in order to improve the quality and readability of the final figures. 

```{r, cache=TRUE, eval=FALSE}
# Create new column called sample_name from the row.name
meta$sample_name=row.names(meta) 
# Separate column sample_name based on hyphen into five new columns 
meta=separate(meta, 
              col=sample_name, 
              into=c("experience", 
                     "short_sample_type", 
                     "time", 
                     "bioreactor", 
                     "replicate"), 
              sep="-") 
# reset the row.name because R 
meta$sample_name=row.names(meta) 
# Replacing abbreviation
meta$experience=gsub("E", "Experience ", meta$experience) # In the column experience, replace E with Experience 
meta$bioreactor=gsub("B0","Aquifer", meta$bioreactor) # In the column bioreactor, replace B00 with Aquifer 
meta$bioreactor=gsub("B", "Bioreactor ",meta$bioreactor) # In the column bioreactor, replace B with Bioreactor 

```

We could then begin the puzzling task of identifying which variables we would use as iterator in our *for loop* to generate the figures and how to present the data on the x axis.
If we go back to our data, we have...  

- 2 experiences ;
- 3 types of samples ;
  - Raw groundwater (identified as *Aquifer*) ;
  - Rock pellets ;
  - Groundwater ;
- 3 bioreactors ;  

... all along a temporal scale. 

We decided that we wanted each figure to represent a different **experience** and **sample type**. We would therefore have four figures :

(1) Experience 1 - Groundwater
(2) Experience 1 - Rock 
(3) Experience 2 - Groundwater 
(4) Experience 2 - Rock 

In each of these figures, we would use `ggplot2` function `facet_grid` to split the graph in different panels based on the values from the column **bioreactor** 
(either *Aquifer*, *Bioreactor 1*, *Bioreactor 2*, *Bioreactor 3*). 

Sampling time would be used for the x-axis. 

Therefore, as iterator, we created a new column called **experience_sample** combining the values from the **experience** and **sample type** column. For the x-axis, we had certain samples collected at the same time to generate different replicates so we couldn't simply use the column **time**  because then those samples would be stacked together. This was notably the case for the aquifer samples from time T00 and all the rock samples from the experience 2. We had to generate a new column which we called **time_replicate** and populated this new column by using the function `ifelse` : 

- if value from the column **bioreactor** = *Aquifer*, then populate column **time_replicate**  with string "*T00.*" (for time 00) followed by the value from the column **replicate**; and 

- if value from the column **experience** = *Experience 2* and **Sample_type** = *Rock*, then populate column **time_replicate**  with the value from the column **time** followed by the string "*R.*" (for replicate) and value from the column **replicate** ; 
- for all other cases (else), simply populate the column **time_replicate** with values from the column **time**. 

```{r, cache=TRUE, eval=FALSE}
# Creating new column for our iterator by combining experience # and sample type 
meta$experience_sample=paste(meta$experience,"-",meta$Sample_type)

# create the new column to be used for the x axis (representing both times and replicate #)
meta$time_replicate=with(meta, 
                         ifelse(bioreactor == "Aquifer", paste("T00.", meta$replicate, sep = ""),
                                ifelse(experience == "Experience 2" & Sample_type =="Rock", 
                                       paste(meta$time,paste("R.", meta$replicate), sep = "-"), meta$time)))

# Getting rid of unnecessary columns
meta=subset(meta, select=-c(position_pastille, temp, ph, do_percent, dic, doc, short_sample_type))
```

This is now the look of our sample metadata table :
```{r, echo=FALSE, cache=TRUE}
library(DT)
datatable(meta, options=list(dom = 'tip', pageLength=5,autoWidth = TRUE,columnDefs=list(list(width="50%", targets=0)),
                                                              initComplete = JS("function(settings, json) {",
                                                                              "$(this.api().table().header()).css({'font-size': '12px'});",
                                                                              "}"))) %>% 
  formatStyle(columns = colnames(.$x$data), `font-size` = '12px')
``` 

## Combine and melt dataframes
We now want to combine into one single dataframe the abundance of the different ASV, the taxonomy assigned to each ASV, and our cleaned up sample metadata dataframe. 
For this we are using the package and function `phyloseq` before melting our dataframe to what we call long format.
```{r, cache=TRUE, eval=FALSE}
# Merge into phyloseq object
ps=phyloseq(otu_table(asv, taxa_are_rows=TRUE), tax_table(as.matrix(taxa)), sample_data(meta))
# Get abundance in %
ps_rel_abund=transform_sample_counts(ps, function(x) x/sum(x)) 
# Agglomerate taxa at taxonomic rank defined earlier 
glom=tax_glom(ps_rel_abund, taxrank = taxa_rank) 
# Melt to long format 
melted_df=psmelt(glom)
```
The look of the melted dataframe : 
```{r, echo=FALSE, warning=FALSE, cache=TRUE}
library(DT)
datatable(melted_df,rownames = FALSE, options=list(dom = 'tip', scrollX=T, pageLength=5,autoWidth = TRUE,
                             initComplete = JS("function(settings, json) {",
                                               "$(this.api().table().header()).css({'font-size': '12px'});",
                                               "}"))) %>%  formatStyle(columns = colnames(.$x$data), `font-size` = '12px')
``` 

## Identify most abundant taxa 
In this step we are using a *for loop* again in order to identify in each sample the X number of most abundant taxa. All other taxa will be renamed as *Others*. 
The X number of taxa to show in every sample can be changed to whatever the user prefers by simply modifying the value for the variable `number_of_taxa`. 
Yet, keep in mind that the most abundant taxa will most probably not be the same in each sample and therefore your legend will likely include more taxa then the number defined here.  
```{r,cache=TRUE, eval=FALSE}
# Defining the number of most abundant taxa to keep
number_of_taxa=5
# Create an empty list that we will populated with the unique taxa of each sample
list_of_all_taxonomic_rank= list() 
i = 0 
# Beginning of the for loop
for (each_sample in unique(melted_df$Sample)){
    i=i+1
    sample=subset(melted_df, Sample==each_sample) # Create a new dataframe from the iterator (sample). 
    total_abundance=aggregate(sample$Abundance, by=list(taxa_rank=sample[[taxa_rank]]), FUN=sum) # Combine together the same taxa and sum the abundances 
    top=head(total_abundance[order(total_abundance$x, decreasing= T),], n=number_of_taxa) # Sort by abundance and keep only the X number of taxa defined by variable number_of_taxa
    others_df=sample[!sample[[taxa_rank]] %in% top$taxa_rank,] # Extract in a new dataframe all taxa that are not present in the dataframe `top`
    others_list=pull(others_df, taxa_rank) # Create a list by pulling all the values from the column corresponding to the taxa_rank into a list
    sample[sample[[taxa_rank]]%in% others_list,][[taxa_rank]]="Others" # In the dataframe `sample` rename all the taxa from the list `others_list` as `Others`
    list_of_all_taxonomic_rank[[i]]=sample #save this dataframe in our list 
}
df=do.call("rbind",list_of_all_taxonomic_rank) # combine all the dataframe from the list into one dataframe
unique_taxon=data.frame(unique(df[[taxa_rank]])) # create dataframe with the unique names of taxa 
name=colnames(unique_taxon) # extract the name of the column in order to rename the column with the following line 
names(unique_taxon)[names(unique_taxon)==name]=as.character(taxa_rank) # Rename the column to the taxa rank defined earlier 
```

## Assign color and italicize taxa 

We are now generating a new set of random unique colors and assigning a specific color to each taxa from our list of most abundant taxa. We are also using the *markdown syntax* to italicize the taxa name in both the legend and dataframe. 
**If you wish to use a previously saved color palette do not run the following chunk of code and jump to section [Saving and loading color range]**
```{r, eval=FALSE, cache=TRUE, eval=FALSE}
# get the total number of unique most abundant taxa 
n=nrow(unique_taxon) 
# generate a set of X unique colors corresponding to the number of unique taxa
palette=distinctColorPalette(n) 
unique_taxon[[taxa_rank]]=factor(unique_taxon[[taxa_rank]])
names(palette)=levels(unique_taxon[[taxa_rank]]) 
# assign gray to category "Others". The same nomenclature can be use to manually change certain colors. 
palette[["Others"]]="#E1E1E1" 

#  recreate palette with markdown to italicize name and remove the underscore after Unclassified
all_names=data.frame(names(palette))
names_markdown=all_names %>%
  mutate(names.palette.=str_replace(names.palette., "(.*)","*\\1*"), # Adding asterisk at beginning and end of every taxa
         names.palette.=str_replace(names.palette., "\\*Unclassified_(.*)\\*","Unclassified *\\1*"), # Removing the asterisk for words that don't need to be italicize (Unclassified and Others)
         names.palette.=str_replace(names.palette., "\\*Others\\*", "Others"))
list_names=as.vector(names_markdown$names.palette.)
# Replace names of object
names(palette)=c(list_names)

# Making the same modification to the taxa name from the legend to the taxa names in the dataframe
df[[taxa_rank]]=str_replace(df[[taxa_rank]], "(.*)","*\\1*")
df[[taxa_rank]]=str_replace(df[[taxa_rank]], "\\*Unclassified_(.*)\\*","Unclassified *\\1*")
df[[taxa_rank]]=str_replace(df[[taxa_rank]], "\\*Others\\*", "Others")

# Ordering the legend in alphabetical order
legend_raw=unique(df[[taxa_rank]]) #Extract legend as text
ordered_legend=sort(legend_raw) # order alphabetically
reordered_legend=fct_relevel(ordered_legend, "Others") # move "Others" to the beginning
final_legend=levels(reordered_legend) # Extract the levels in a new object

my_scale <- scale_fill_manual(name=as.character(taxa_rank), breaks=paste(final_legend), values = palette, na.translate=FALSE, drop=TRUE, limits = force) # Recreate color fill
```

## Saving and loading color range 

If you wish to save your current color palette use the following line of code to save it as an R data file. 
```{r, eval=FALSE}
# Create a name for saving the object which includes both the rank and number of taxa
scale_name=paste(taxa_rank,"-",number_of_taxa,"-","colors",".rds", sep="")
# Save palette using the defined name 
saveRDS(my_scale,as.character(scale_name)) 
```
If you wish to load a previously saved color palette use the following line of code. 
```{r, eval=FALSE}
my_scale=readRDS("name_of_your_palette.rds") # Load color palette
```

## Plotting your figures
Finally this is where we are generating the final figures. As you can see we are again using a *for loop* to generate multiple figures with only one chunk of code. 
As a reminder we decided to generate **four figures** based on the experience number and sample type. Hence we generated a new column called **experience_sample** which we are using here as our iterator. 
We are using the function `fct_reorder` to align along the bottom and top axis the most abundant taxa. 
```{r, cache=TRUE, eval=FALSE}
plot_list = list() # Generate empty list which we will populate with our generated figures
i = 0

for (experiences_and_sample in unique(df$experience_sample)){ 
  i = i + 1
  df_sample=subset(df, experience_sample==experiences_and_sample)
  p=ggplot(df_sample, aes(x=time_replicate, weight=Abundance, fill=fct_reorder(.data[[taxa_rank]],Abundance,.desc=FALSE))) + # .data is very important to force the evaluation of the input variables (taxonomic_rank)
  geom_bar() +
  labs(y ='Relative abundance (%)', x="Time") +
  scale_y_continuous(expand = c(0,0)) + # Remove the white space 
  theme_classic() +
  theme(text = element_text(size = 9),
        axis.title=element_text(size=8),
        legend.title=element_text(size=8),
        plot.title = element_text(hjust =0.5),
        axis.text.x = element_text(angle=70,vjust = 0.5, size=7),
        legend.position="bottom", 
        legend.text=element_markdown(size=7),
        legend.key.size = unit(0.5, 'cm'), # adjusting the size of the legend colored box
        strip.background = element_blank()) + # remove facet_grid box background 
  ggtitle(paste(experiences_and_sample)) + # Use the value from our new column as plot title. This title will then be extracted to save the figures. 
  my_scale + # Load our color palette 
  facet_grid(~bioreactor, scales="free", space="free") + # facet_grid according to bioreactor 
  guides(fill=guide_legend(nrow=6, title=taxa_rank, title.position="top", title.hjust = 0.5, reverse=FALSE)) # Adjusting the legend, notably the number of rows and position
  plot_list[[i]] = p # save plot into list 
}
# Uncomment the following line to visualize your plots in R. 
# plot_list 

# Set path for saving plots
plot_path="figures" # Set the path to where you want to save your figures
for (i in plot_list[]){
  title=extractGGTitle(i) # extract title from plot
  titlepdf=paste(taxa_rank,"-", title,".pdf", sep="") # add taxa rank before title and .pdf to the end (to save figure as pdf)
  ggsave(as.character(titlepdf),plot=i, device="pdf", path=plot_path, width=27.94, height = 21.59, units="cm") # save plot
}
```

## Export tables
Now that we have generated our nice figures, I also recommend exporting the table as a comma separated value (CSV) file. By doing so, you will be able to access the numbers behind the graph, which can be interesting data to present in the **Results** section of your article.

We are first generating a new column called **Relative_abundance** which combines all the variables (**experience_sample**, **time_replicate**, and **taxa_rank**) used in generating the figures. 
We then sum the relative abundance of similar taxa with the function `aggregate`.
```{r, eval=FALSE}
df$Relative_abundance=(paste(df$experience_sample, "_", df$time_replicate,"_", df[[taxa_rank]]))
supplementary_dataframe=aggregate(x=df$Abundance, by=list(df$Relative_abundance), FUN=sum)

# Save tables
title_df=paste(taxa_rank,"-","supplementary_dataframe",".csv", sep="") 
write.csv(supplementary_dataframe, as.character(title_df), quote=FALSE, row.names=FALSE)
```

## Complete code 

You can copy-paste the following block of code inside a new markdown document. 
Code-chunks will be automatically generated and you can use the far right button (<font color='green'> ▶ </font>) to execute all of the code inside each chunk.

````{verbatim, lang = "python"}
```{r}
# ----------- Load libraries -----------
library(devtools)
library(ggTag) # To extract title from plots
library(forcats) # To reorder our factors (function "fct_relevel")
library(dplyr) # Dataframe manipulation (notably function "pull")
library(tidyr) # Dataframe manipulation (function "separate")
library(phyloseq) # Very pratical library for the analysis of amplicon data 
library(randomcoloR) # Generate sets of random colors
library(ggplot2) # Generate plots 
library(stringr) # Makes working with strings as easy as possible (function "str_replace")
library(ggtext) # Allows the use of markdown text (used to make names italic)
library(glue)

# ----------- Define path -----------
path="~/project/domain/int_data"

asv=read.table(file = glue("{path}/rarefied_asv_bac.csv"), sep=",", row.names=1, header=TRUE, check.names=FALSE)
taxa=read.table(file = glue("{path}/rarefied_taxa_bac.csv"), sep=",", row.names=1, header=TRUE)
meta=read.table(file = glue("{path}/rarefied_meta_bac.csv"), sep=",", row.names=1, header=TRUE)

# ----------- Define rank of interest -----------
taxa_rank="Genus"

# ----------- Merge into phyloseq object -----------
ps=phyloseq(otu_table(asv, taxa_are_rows=TRUE), tax_table(as.matrix(taxa)), sample_data(meta))
# Get abundance in %
ps_rel_abund=transform_sample_counts(ps, function(x) x/sum(x)) 
# Agglomerate taxa at taxonomic rank defined earlier 
glom=tax_glom(ps_rel_abund, taxrank = taxa_rank) 
# Melt to long format 
melted_df=psmelt(glom)

# Defining the number of most abundant taxa to keep
number_of_taxa=5

# ----------- Identifying most abundant taxa per sample -----------

# Create an empty list that we will populated with the unique taxa of each sample
list_of_all_taxonomic_rank= list() 
i = 0 
# Beginning of the for loop
for (each_sample in unique(melted_df$Sample)){
    i=i+1
    sample=subset(melted_df, Sample==each_sample) # Create a new dataframe from the iterator (sample). 
    total_abundance=aggregate(sample$Abundance, by=list(taxa_rank=sample[[taxa_rank]]), FUN=sum) # Combine together the same taxa and sum the abundances 
    top=head(total_abundance[order(total_abundance$x, decreasing= T),], n=number_of_taxa) # Sort by abundance and keep only the X number of taxa defined by variable number_of_taxa
    others_df=sample[!sample[[taxa_rank]] %in% top$taxa_rank,] # Extract in a new dataframe all taxa that are not present in the dataframe `top`
    others_list=pull(others_df, taxa_rank) # Create a list by pulling all the values from the column corresponding to the taxa_rank into a list
    sample[sample[[taxa_rank]]%in% others_list,][[taxa_rank]]="Others" # In the dataframe `sample` rename all the taxa from the list `others_list` as `Others`
    list_of_all_taxonomic_rank[[i]]=sample #save this dataframe in our list 
}
df=do.call("rbind",list_of_all_taxonomic_rank) # combine all the dataframe from the list into one dataframe
unique_taxon=data.frame(unique(df[[taxa_rank]])) # create dataframe with the unique names of taxa 
name=colnames(unique_taxon) # extract the name of the column in order to rename the column with the following line 
names(unique_taxon)[names(unique_taxon)==name]=as.character(taxa_rank) # Rename the column to the taxa rank defined earlier 

# ---- Create color range and italicize the taxa name ---- 
n=nrow(unique_taxon) # get the total number of unique most abundant taxa 
palette=distinctColorPalette(n) # generate a set of X unique colors corresponding to the number of unique taxa
unique_taxon[[taxa_rank]]=factor(unique_taxon[[taxa_rank]])
names(palette)=levels(unique_taxon[[taxa_rank]]) # 
palette[["Others"]]="#E1E1E1" # assign gray to category "Others". The same nomenclature can be use to manually change certain colors. 

#  recreate palette with markdown to italicize name and remove the underscore after Unclassified
all_names=data.frame(names(palette))
names_markdown=all_names %>%
  mutate(names.palette.=str_replace(names.palette., "(.*)","*\\1*"), # Adding asterisk at beginning and end of every taxa
         names.palette.=str_replace(names.palette., "\\*Unclassified_(.*)\\*","Unclassified *\\1*"), # Remove asterisk for words that don't need to be italicize (Unclassified and Others)
         names.palette.=str_replace(names.palette., "\\*Others\\*", "Others"))
list_names=as.vector(names_markdown$names.palette.)
# Replace names of object
names(palette)=c(list_names)
# Make same modification to the taxa name from the legend to the taxa names in the dataframe
df[[taxa_rank]]=str_replace(df[[taxa_rank]], "(.*)","*\\1*")
df[[taxa_rank]]=str_replace(df[[taxa_rank]], "\\*Unclassified_(.*)\\*","Unclassified *\\1*")
df[[taxa_rank]]=str_replace(df[[taxa_rank]], "\\*Others\\*", "Others")
# Ordering the legend in alphabetical order
legend_raw=unique(df[[taxa_rank]]) #Extract legend as text
ordered_legend=sort(legend_raw) # order alphabetically
reordered_legend=fct_relevel(ordered_legend, "Others") # move "Others" to the beginning
final_legend=levels(reordered_legend) # Extract the levels in a new object
# Recreate color fill
my_scale <- scale_fill_manual(name=as.character(taxa_rank), breaks=paste(final_legend), values = palette, na.translate=FALSE, drop=TRUE, limits = force) 

# ---- Plot figures ---- 
plot_list = list() # Generate empty list which we will populate with our generated figures
i = 0

for (experiences_and_sample in unique(df$experience_sample)){ 
  i = i + 1
  df_sample=subset(df, experience_sample==experiences_and_sample)
  p=ggplot(df_sample, aes(x=time_replicate, weight=Abundance, fill=fct_reorder(.data[[taxa_rank]],Abundance,.desc=FALSE))) + # .data is very important to force the evaluation of the input variables (taxonomic_rank)
  geom_bar() +
  labs(y ='Relative abundance (%)', x="Time") +
  scale_y_continuous(expand = c(0,0)) + # Remove the white space 
  theme_classic() +
  theme(text = element_text(size = 9),
        axis.title=element_text(size=8),
        legend.title=element_text(size=8),
        plot.title = element_text(hjust =0.5),
        axis.text.x = element_text(angle=70,vjust = 0.5, size=7),
        legend.position="bottom", 
        legend.text=element_markdown(size=7),
        legend.key.size = unit(0.5, 'cm'), # adjusting the size of the legend colored box
        strip.background = element_blank()) + # remove facet_grid box background 
  ggtitle(paste(experiences_and_sample)) + # Use the value from our new column as plot title. This title will then be extracted to save the figures. 
  my_scale + # Load our color palette 
  facet_grid(~bioreactor, scales="free", space="free") + # facet_grid according to bioreactor 
  guides(fill=guide_legend(nrow=6, title=taxa_rank, title.position="top", title.hjust = 0.5, reverse=FALSE)) # Adjusting the legend, notably the number of rows and position
  plot_list[[i]] = p # save plot into list 
}
``` 

````
