<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Example workflow of amplicon sequence data analysis</title>
  <meta name="description" content="Example workflow of amplicon sequence data analysis" />
  <meta name="generator" content="bookdown 0.35 and GitBook 2.6.7" />

  <meta property="og:title" content="Example workflow of amplicon sequence data analysis" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Example workflow of amplicon sequence data analysis" />
  
  
  

<meta name="author" content="Karine Villeneuve" />


<meta name="date" content="2024-09-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="dada2-pipeline.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="toc.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Amplicon sequencing data analysis</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#setting-up-your-environment"><i class="fa fa-check"></i><b>1.1</b> Setting up your environment</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#download-fastqs"><i class="fa fa-check"></i><b>1.2</b> Download fastqs</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html"><i class="fa fa-check"></i><b>2</b> DADA2 pipeline</a>
<ul>
<li class="chapter" data-level="2.1" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#general-workflow"><i class="fa fa-check"></i><b>2.1</b> General workflow</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#getting-ready"><i class="fa fa-check"></i><b>2.1.1</b> Getting ready</a></li>
<li class="chapter" data-level="2.1.2" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#inspect-quality"><i class="fa fa-check"></i><b>2.1.2</b> Inspect quality</a></li>
<li class="chapter" data-level="2.1.3" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#filter-and-trim-sequences"><i class="fa fa-check"></i><b>2.1.3</b> Filter and trim sequences</a></li>
<li class="chapter" data-level="2.1.4" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#learn-error-rates"><i class="fa fa-check"></i><b>2.1.4</b> Learn error rates</a></li>
<li class="chapter" data-level="2.1.5" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#merge-paired-reads"><i class="fa fa-check"></i><b>2.1.5</b> Merge paired reads</a></li>
<li class="chapter" data-level="2.1.6" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#track-reads-through-pipeline"><i class="fa fa-check"></i><b>2.1.6</b> Track reads through pipeline</a></li>
<li class="chapter" data-level="2.1.7" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#classify-sequences"><i class="fa fa-check"></i><b>2.1.7</b> Classify sequences</a></li>
<li class="chapter" data-level="2.1.8" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#add-highest-identified-taxonomic-rank-to-unclassified-ranks"><i class="fa fa-check"></i><b>2.1.8</b> Add highest identified taxonomic rank to unclassified ranks</a></li>
<li class="chapter" data-level="2.1.9" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#removing-contaminants"><i class="fa fa-check"></i><b>2.1.9</b> Removing contaminants</a></li>
<li class="chapter" data-level="2.1.10" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#generate-a-phylogenetic-tree"><i class="fa fa-check"></i><b>2.1.10</b> Generate a phylogenetic tree</a></li>
<li class="chapter" data-level="2.1.11" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#add-dna-sequences"><i class="fa fa-check"></i><b>2.1.11</b> Add DNA sequences</a></li>
<li class="chapter" data-level="2.1.12" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#shorten-asv-name"><i class="fa fa-check"></i><b>2.1.12</b> Shorten ASV name</a></li>
<li class="chapter" data-level="2.1.13" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#save-tables"><i class="fa fa-check"></i><b>2.1.13</b> Save tables</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#complete-code"><i class="fa fa-check"></i><b>2.2</b> Complete code</a></li>
<li class="chapter" data-level="2.3" data-path="dada2-pipeline.html"><a href="dada2-pipeline.html#forward-reads-only"><i class="fa fa-check"></i><b>2.3</b> Forward reads only</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="rarefaction.html"><a href="rarefaction.html"><i class="fa fa-check"></i><b>3</b> Rarefaction</a>
<ul>
<li class="chapter" data-level="3.1" data-path="rarefaction.html"><a href="rarefaction.html#load-required-libraries"><i class="fa fa-check"></i><b>3.1</b> Load required libraries</a></li>
<li class="chapter" data-level="3.2" data-path="rarefaction.html"><a href="rarefaction.html#getting-ready-1"><i class="fa fa-check"></i><b>3.2</b> Getting ready</a></li>
<li class="chapter" data-level="3.3" data-path="rarefaction.html"><a href="rarefaction.html#filter-low-abundant-taxa"><i class="fa fa-check"></i><b>3.3</b> Filter low abundant taxa</a></li>
<li class="chapter" data-level="3.4" data-path="rarefaction.html"><a href="rarefaction.html#generate-rarefaction-curve"><i class="fa fa-check"></i><b>3.4</b> Generate rarefaction curve</a></li>
<li class="chapter" data-level="3.5" data-path="rarefaction.html"><a href="rarefaction.html#customizing-the-graph"><i class="fa fa-check"></i><b>3.5</b> Customizing the graph</a></li>
<li class="chapter" data-level="3.6" data-path="rarefaction.html"><a href="rarefaction.html#comparing-richness-and-diversity-between-rarefied-and-un-rarefied-samples"><i class="fa fa-check"></i><b>3.6</b> Comparing richness and diversity between rarefied and un-rarefied samples</a></li>
<li class="chapter" data-level="3.7" data-path="rarefaction.html"><a href="rarefaction.html#complete-code-1"><i class="fa fa-check"></i><b>3.7</b> Complete code</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Example workflow of amplicon sequence data analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Example workflow of amplicon sequence data analysis</h1>
<p class="author"><em>Karine Villeneuve</em></p>
<p class="date"><em>2024-09-17</em></p>
</div>
<div id="introduction" class="section level1 hasAnchor" number="1">
<h1><span class="header-section-number">1</span> Introduction<a href="index.html#introduction" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>This guide is intended to walk the user through the typical R workflow for processing raw amplicon sequencing data from paired end Illumina Miseq data into a table of exact amplicon sequence variants (ASVs) present in each sample. Please note that many other extensive documentation and tutorial pages are available for packages used in this workflow, notably <a href="https://benjjneb.github.io/dada2/tutorial.html">DADA2</a> and <a href="https://joey711.github.io/phyloseq/">phyloseq</a>. Therefore, for issues or for further questions about certain functions I always recommend consulting the available R documentation and if available the relevant github issues sites for answers.</p>
<p>This guide assumes the following :</p>
<ol style="list-style-type: decimal">
<li><p>The user has access to one of the server from the Lazar lab (Orion, Hercules, Ulysse).</p></li>
<li><p>The user has an active VPN access.</p></li>
<li><p>The paired-end fastq files from Illumina Miseq sequencing were transferred to the user’s home directory of her/his server. Please see section <a href="index.html#download-fastqs">Download fastqs</a> for more details.</p></li>
<li><p>The user has followed the <code>Introduction to linux</code> guide (available on the lab’s TEAM) and is comfortable with basic command line functions such as :</p>
<ul>
<li>listing files inside a current directory (<code>ls</code>) ;</li>
<li>moving from one directory to the other (<code>cd</code>) ;</li>
<li>creating new directory (<code>mkdir</code>) ; and</li>
<li>moving / copying (<code>mv</code>/<code>cp</code>) files from one directory to the other.</li>
</ul></li>
</ol>
<div id="setting-up-your-environment" class="section level2 hasAnchor" number="1.1">
<h2><span class="header-section-number">1.1</span> Setting up your environment<a href="index.html#setting-up-your-environment" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Keeping your files organized is a skill that has a high long-term payoff. As you are in the thick of an analysis, you may underestimate how many files/folders you have floating around. But a short time later, you may return to your files and realize your organization was not as clear as you hoped, which can ultimately lead to significantly slower research progress. Furthermore, one must keep in mind that someone unfamiliar with your project should be able to look at your computer files and understand in detail what you did and why.</p>
<p>While there’s a lot of ways to keep your files organized, and there’s not a “one size fits all” organizational solution, below we propose a simple organizational scheme which is project-oriented, maintainable and ultimately follows consistent patterns for amplicon sequence analysis. Please note that the proposed workflow assumes such organization.</p>
<pre><code>┌─ ~ -------------------------------- Your home directory
│   ├── chapter1_16S_diversity ------ Project with a short but meaningfull name with  
│       │                             a second level domain-specific organization (if applicable). 
│       ├── archaea 
│       ├── eukaryotes
│       ├── bacteria ---------------- Third level domain-specific organization :
│           ├── data   -------------- final datasets generated by scripts to be used for interpretation ;
│           ├── int_data  ----------- intermediate data to be used by other script (DADA2 output, rarefied data, etc.) ;
│           ├── figures  ------------ figures generated by scripts ; 
│           ├── raw_data   ---------- raw files (i.e. fastq files generated by sequencing, raw metadata table, etc.) ; and
│                 ├── BAC_sample-1_R1.fastq.gz
│                 ├── BAC_sample-1_R2.fastq.gz
│                 ├── BAC_sample-2_R1.fastq.gz
│                 ├── BAC_sample-2_R2.fastq.gz
│                 └── ...
│           └── scripts   ----------- executable and scripts. 
│                  ├── bac_DADA2.rmd
│                  ├── bac_rarefy.rmd
│                  ├── bac_alpha_div.rmd
│                  ├── bac_stacked_barchart.rmd
│                  └── ... 
│       ├── data   ------------------ Project-specific final datasets for publication
│       ├── figures ----------------- Project-specific final figures for publication
│       └── scripts ----------------- Project-specific scripts (only if applicable, i.e. multi-domain, explanatory variables, etc.)
└──────────────────────────────────────────────────────────────────</code></pre>
<p>Further reading about organizing files and foldes :</p>
<ul>
<li><p><a href="https://jhudatascience.org/Reproducibility_in_Cancer_Informatics/organizing-your-project.html">Organizing your project</a> by the Johns Hopkins Data Science Lab</p></li>
<li><p><a href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424#s9">A Quick Guide to Organizing Computational Biology Projects</a> by William Stafford Noble, 2009</p></li>
<li><p><a href="https://www.reddit.com/r/bioinformatics/comments/3rdlhf/help_how_do_you_organize_your_projectsfiles/">Reddit post</a></p></li>
<li><p><a href="https://www.mdc-berlin.de/research-data-management/organizing-your-data">Organizing your data</a> by The Max Delbrück Center</p></li>
</ul>
</div>
<div id="download-fastqs" class="section level2 hasAnchor" number="1.2">
<h2><span class="header-section-number">1.2</span> Download fastqs<a href="index.html#download-fastqs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>How to download fastqs files from Illumina BaseSpace Sequence Hub</p>
<ol style="list-style-type: decimal">
<li><p>Open the link found in the email sent by Geneviève Bourret.</p>
<p>If this is your first time downloading your fastqs create a new BaseSpace Sequence Hub account using the email address to which the email from Geneviève was addressed (normally this would be your UQAM’s email)</p></li>
<li><p>On the pop-up window informing you that the CERMO-FC has shared the following item with you click <code>ACCEPT</code>.</p></li>
<li><p>Click on the <code>PROJECTS</code> tab in the upper section of the page.</p></li>
<li><p>Select your project and then click on the second round logo from the left which looks like a blank page and in the drop-down menu select <code>DOWNLOAD</code> then <code>PROJECT</code>.</p></li>
<li><p>If required, download the Illumina Basespace downloader by clicking <code>INSTALL DOWNLOAD</code> and follow the instructions. Otherwise simply click <code>DOWNLOAD</code> to begin downloading your fastqs.</p></li>
<li><p>Once the download is complete you will find inside the folder a folder for each of your sample inside which the forward and reverse read are both found in another folder. Instead of going into each folder individually and copying the fastqs manually we can use the terminal to do the job for us. From a new local terminal window navigate to the folder containing all the folders and execute the following command after having modified <code>/path/to/directory/where/to/move/fastqs</code> to the actual path where you wish to move your fastqs.</p></li>
</ol>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="index.html#cb2-1" tabindex="-1"></a><span class="fu">find</span> ./ <span class="at">-name</span> <span class="st">&quot;*.gz&quot;</span> <span class="at">-exec</span> cp <span class="at">-prv</span> <span class="st">&quot;{}&quot;</span> <span class="st">&quot;/path/to/directory/where/to/move/fastqs&quot;</span> <span class="st">&quot;;&quot;</span></span></code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li><p>Finally you can transfer your fastqs to your assigned server using any File Transfer Protocol (FTP) clients (such as <a href="https://filezilla-project.org/">FileZilla</a> or <a href="https://cyberduck.io/">Cyberduck</a>) or using the SCP (secure copy) command-line utility.</p>
<p>For SCP you can copy an entire folder by opening a new local terminal window and navigating to the directory containing the folder with the fastqs. From that directory execute the following command. You will then be asked to enter the password for your user on the server.</p></li>
</ol>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="index.html#cb3-1" tabindex="-1"></a><span class="fu">scp</span> <span class="at">-r</span> name_of_foler_with_fastqs username@server.bio.uqam.ca:/path/to/copy/folder</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="dada2-pipeline.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": null,
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/index.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
